### PR TITLE
Remove `ipython_genutils` as a dependency

### PR DESCRIPTION
**Pull request by Carreau**
This vendors the code. 

~~There is still 1 place using ipython_genutils (`ensure_dir_exists`), but basically it is only called if you call `export_html(..., inline=False)`, which I can't find anywhere. So if you are willing to remove support for html export with non-inline html, then you can drop ipython_genutils dependency.~~

### CODE REVIEW

File: qtconsole/util.py [hunk 1]
```diff
@@ -106,3 +106,215 @@ def get_font(family, fallback=None):
...
+
+    Parameters
+    ----------
+    items
+        list of strings to columize
+    row_first : (default False)
+        Whether to compute columns for a row-first matrix instead of
+        column-first (default).
+    empty : (default None)
+        default value to fill list if needed
```
Comment (ccordoba12) [inline qtconsole/util.py:162]:
```suggestion
        Default value to fill list if needed
```

File: qtconsole/util.py [hunk 2]
```diff
@@ -106,3 +106,215 @@ def get_font(family, fallback=None):
...
+    separator_size : int (default=2)
+        How much characters will be used as a separation between each columns.
```
Comment (ccordoba12) [inline qtconsole/util.py:164]:
```suggestion
        How much characters will be used as a separation between each column.
```

File: qtconsole/util.py [hunk 3]
```diff
@@ -106,3 +106,215 @@ def get_font(family, fallback=None):
...
+    displaywidth : int (default=80)
+        The width of the area onto which the columns should enter
+
+    Returns
+    -------
+    strings_matrix
+        nested list of string, the outer most list contains as many list as
```
Comment (ccordoba12) [inline qtconsole/util.py:171]:
```suggestion
        Nested list of strings, the outer most list contains as many list as
```

File: qtconsole/util.py [hunk 4]
```diff
@@ -106,3 +106,215 @@ def get_font(family, fallback=None):
...
+        rows, the innermost lists have each as many element as columns. If the
+        total number of elements in `items` does not equal the product of
+        rows*columns, the last element of some lists are filled with `None`.
+    dict_info
+        some info to make columnize easier:
```
Comment (ccordoba12) [inline qtconsole/util.py:176]:
```suggestion
        Some info to make columnize easier:
```

File: qtconsole/util.py [hunk 5]
```diff
@@ -106,3 +106,215 @@ def get_font(family, fallback=None):
...
+    displaywidth : int, optional [default is 80]
+        Width of the display in number of characters.
+
+    Returns
+    -------
+    The formatted string.
+    """
+    if not items:
+        return "\n"
+    matrix, info = compute_item_matrix(
```
Comment (ccordoba12) [inline qtconsole/util.py:246]:
```suggestion

    matrix, info = compute_item_matrix(
```

File: qtconsole/util.py [hunk 6]
```diff
@@ -106,3 +106,215 @@ def get_font(family, fallback=None):
...
+        items,
+        row_first=row_first,
+        separator_size=len(separator),
+        displaywidth=displaywidth,
+    )
+    if spread:
```
Comment (ccordoba12) [inline qtconsole/util.py:252]:
```suggestion

    if spread:
```

File: qtconsole/util.py [hunk 7]
```diff
@@ -106,3 +106,215 @@ def get_font(family, fallback=None):
...
+        separator = separator.ljust(int(info["optimal_separator_width"]))
+    fmatrix = [filter(None, x) for x in matrix]
+    sjoin = lambda x: separator.join(
+        [y.ljust(w, " ") for y, w in zip(x, info["column_widths"])]
+    )
+    return "\n".join(map(sjoin, fmatrix)) + "\n"
```
Comment (ccordoba12) [inline qtconsole/util.py:258]:
```suggestion

    return "\n".join(map(sjoin, fmatrix)) + "\n"
```

File: qtconsole/util.py [hunk 8]
```diff
@@ -106,3 +106,215 @@ def get_font(family, fallback=None):
...
+    >>> get_text_list(['a', 'b', 'c', 'd'], " = ", sep=" + ")
+    'a + b + c = d'
+    """
+    if len(list_) == 0:
+        return ""
+    if wrap_item_with:
+        list_ = ["%s%s%s" % (wrap_item_with, item, wrap_item_with) for item in list_]
+    if len(list_) == 1:
+        return list_[0]
+    return "%s%s%s" % (sep.join(i for i in list_[:-1]), last_sep, list_[-1])
```
Comment (ccordoba12) [inline qtconsole/util.py:288]:
```suggestion

    return "%s%s%s" % (sep.join(i for i in list_[:-1]), last_sep, list_[-1])
```

File: qtconsole/util.py [hunk 9]
```diff
@@ -106,3 +106,215 @@ def get_font(family, fallback=None):
...
+    name : string
+      The fully qualified name of the module/package being imported.
+
+    Returns
+    -------
+    mod : module object
+       The module that was imported.
+    """
+
+    parts = name.rsplit(".", 1)
```
Comment (ccordoba12) [inline qtconsole/util.py:308]:
```suggestion
    parts = name.rsplit(".", 1)

```

**Review by ccordoba12 — COMMENTED** [review]
Thanks @Carreau for your help with this! Some minor comments, the rest looks good to me.

**Comment by ccordoba12** [comment]
> So if you are willing to remove support for html export with non-inline html, then you can drop ipython_genutils dependency.

I agree with that. Could you do it in this PR as well?

**Comment by Carreau** [comment]
yes I can do that early next week.

**Comment by ccordoba12** [comment]
Great! Thanks @Carreau!

File: qtconsole/util.py [hunk 10]
```diff
@@ -106,3 +106,215 @@ def get_font(family, fallback=None):
...
+                    _get_or_default(items, c * nrow + r, default=empty)
+                    for c in range(ncol)
+                ]
+                for r in range(nrow)
+            ],
+            info,
+        )
+
+
+def columnize(items, row_first=False, separator="  ", displaywidth=80, spread=False):
```
Comment (Carreau) [inline qtconsole/util.py:225]: Side question, `row_first=` and `spread=` are never used in qtconsole code. DO you want me to also remove them and simplify everything ?

File: qtconsole/util.py (hunk 8 above)
Comment (Carreau) [inline qtconsole/util.py:288]: Also I got overzealous vendoring I think this one is not necessary.

File: qtconsole/util.py (hunk 10 above)
Comment (ccordoba12) [reply qtconsole/util.py:225]: Sure, that's better.

File: qtconsole/util.py (hunk 8 above)
Comment (ccordoba12) [reply qtconsole/util.py:288]: Ok, then please remove it.

File: qtconsole/console_widget.py [hunk 11]
```diff
@@ -1680,28 +1680,6 @@ def _flush_pending_stream(self):
...
-        -------
-        The formatted string.
-        """
-        # Calculate the number of characters available.
-        width = self._control.document().textWidth()
-        char_width = self._get_font_width()
-        displaywidth = max(10, (width / char_width) - 1)
-
-        return columnize(items, separator, displaywidth)
-
```
Comment (Carreau) [inline qtconsole/console_widget.py:1704]: For git-log of qtconsole, last usage of that was 2012, and is the last place where 'seprator' of columnize is configurable.

File: qtconsole/rich_jupyter_widget.py [hunk 12]
```diff
@@ -22,6 +21,23 @@
...
+    """
+    if not os.path.exists(path):
+        try:
+            os.makedirs(path, mode=mode)
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise
+    elif not os.path.isdir(path):
+        raise IOError("%r exists but is not a directory" % path)
+
```
Comment (Carreau) [inline qtconsole/rich_jupyter_widget.py:40]: I was wrong about export_html(,.... inline=...) being unused, it is still hooked to an option in the UI. So I vendored `_ensure_dir_exists` and made it private to this file.

File: qtconsole/util.py [hunk 13]
```diff
@@ -106,3 +106,158 @@ def get_font(family, fallback=None):
...
+    else:
+        return mylist[i]
+
+
+def compute_item_matrix(items, empty=None, *, separator_size=2, displaywith=80):
+    """Returns a nested list, and info to columnize items
+
+    Parameters
+    ----------
+
```
Comment (Carreau) [inline qtconsole/util.py:149]:
```suggestion
```

**Comment by Carreau** [comment]
Going through my list of opened PRs, is there anything more I can do here ?

Thanks !

**Comment by ccordoba12** [comment]
> > So if you are willing to remove support for html export with non-inline html, then you can drop ipython_genutils dependency.
>
> I agree with that. Could you do it in this PR as well?

I think this is still unaddressed. I mean, I thought you were going to remove our dependency on `ipython_genutils` as part of this PR.

**Comment by Carreau** [comment]
Oh, yeah sorry, I forgot to remove the lines in setup.py and environment.yml, sorry, oversight. 

For html export i was wrong about export_html(,.... inline=...) being unused, so I vendored _ensure_dir_exists. Though the dependency should now be gone.

**Review by ccordoba12 — APPROVED** [review]
Thanks for your help with this @Carreau!

**Pull request merged**